### PR TITLE
optional data parameter in delete methods

### PR DESCRIPTION
**Pull request by bkotyik**
I added the ability to pass request body parameter to DELETE requests. Also added two tests and modified the readme according to the changes.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 node_modules
 bower_components
 dist/restful.js
+.idea
```
Comment (RobinBressan) [inline .gitignore]: Should be in your global `.gitignore`

File: src/model/endpoint.js [hunk 2]
```diff
@@ -202,8 +202,8 @@ export default function endpoint(url, parent) {
             );
         },
 
-        delete(headers) {
-            var nextConfig = _generateRequestConfig('delete', url, {}, headers);
+        delete(headers,data) {
+            var nextConfig = _generateRequestConfig('delete', url, {}, headers,data);
```
Comment (RobinBressan) [inline src/model/endpoint.js]: CS

File: src/model/member.js [hunk 3]
```diff
@@ -53,8 +53,8 @@ export default function member(name, id, parent) {
                 });
         },
 
-        delete(headers) {
-            return refEndpoint.delete(headers)
+        delete(headers,data) {
```
Comment (RobinBressan) [inline src/model/member.js]: CS

File: src/restful.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ export default function restful(baseUrl, port) {
         baseUrl: baseUrl,
         port: port || 80,
         prefixUrl: '',
-        protocol: 'http',
+        protocol: 'http'
```
Comment (RobinBressan) [inline src/restful.js]: Same as above, you must keep it.

File: src/model/member.js [hunk 5]
```diff
@@ -53,8 +53,8 @@ export default function member(name, id, parent) {
...
+            return refEndpoint.delete(headers,data)
```
Comment (RobinBressan) [inline src/model/member.js]: CS

File: src/restful.js [hunk 6]
```diff
@@ -21,7 +21,7 @@ export default function restful(baseUrl, port) {
             fullRequestInterceptors: [],
             fullResponseInterceptors: [],
             requestInterceptors: [],
-            responseInterceptors: [],
+            responseInterceptors: []
```
Comment (RobinBressan) [inline src/restful.js]: same

File: src/model/endpoint.js [hunk 7]
```diff
@@ -202,8 +202,8 @@ export default function endpoint(url, parent) {
             );
         },
 
-        delete(headers) {
-            var nextConfig = _generateRequestConfig('delete', url, {}, headers);
+        delete(headers,data) {
```
Comment (RobinBressan) [inline src/model/endpoint.js]: CS

File: src/model/collection.js [hunk 8]
```diff
@@ -84,9 +84,9 @@ export default function collection(name, parent) {
                 });
         },
 
-        delete(id, headers) {
+        delete(id, headers, data) {
             return refEndpointFactory(id)
-                .delete(headers)
+                .delete(headers,data)
```
Comment (RobinBressan) [inline src/model/collection.js]: CS: Add a space before data

File: src/model/collection.js [hunk 9]
```diff
@@ -94,7 +94,7 @@ export default function collection(name, parent) {
 
         url() {
             return url;
-        },
+        }
```
Comment (RobinBressan) [inline src/model/collection.js]: You must keep this, to avoid useless diff when adding some methods.

File: src/model/member.js (hunk 3 above)
Comment (RobinBressan) [inline src/model/member.js]: CS

File: test/src/restful-spec.js [hunk 10]
```diff
@@ -1012,5 +1012,41 @@
...
+                fullResponseInterceptors: [],
+                transformResponse: [jasmine.any(Function)]
+            });
+        });
+
+        it('should call delete not only with header but even with data when deleting a collection entity', function() {
+            var articles = resource.allUrl('articles', 'https://localhost:3000/v1/articles');
+            spyOn(httpBackend, 'delete').andCallThrough();
+
+            articles.delete(1,{},{key:"value"});
```
Comment (RobinBressan) [inline test/src/restful-spec.js]: use simple quote

Comment (RobinBressan) [inline test/src/restful-spec.js]: add spaces for better readability: `{ key: 'value' }`

File: test/src/restful-spec.js [hunk 11]
```diff
@@ -1012,5 +1012,41 @@
...
+
+            expect(httpBackend.delete).toHaveBeenCalledWith({
+                method: 'delete',
+                url: 'https://localhost:3000/v1/articles/1',
+                params: {},
+                headers: { },
+                data: {key:"value"},
```
Comment (RobinBressan) [inline test/src/restful-spec.js]: same

**Comment by RobinBressan** [comment]
I am wondering if we should put the `data` attribute for the `delete` method before `headers` one as it is done in the `post` `put` `patch` even if it is optional? ping @jpetitcolas @jeromemacias

**Comment by bkotyik** [comment]
Thank you for your review. I hopefully managed to correct my code according to review. 
Changing the order of the attributes wouldnt break the backward compatibility?

**Comment by jeromemacias** [comment]
I think we should set `data` before `headers`:
* for `member`: `delete ( [ data, headers ] )`
* for `collection`: `delete ( id [, data, headers ] )`

But it breaks BC.
Also we will need to add `params` arg in another PR (see #22), and so it should be `delete ( [params, data, headers ] )`. This will break BC again...

**Comment by bkotyik** [comment]
Done! Waiting for your review.

File: README.md [hunk 12]
```diff
@@ -201,7 +201,7 @@ Restful.js exposes similar methods on collections, members and entities. The nam
 * `get ( id [, params [, headers ]] )`: Get a member in a collection. Returns a promise with an entity.
 * `post ( data [, headers ] )`: Create a member in a collection. Returns a promise with the response.
 * `put ( id, data [, headers ] )`: Update a member in a collection. Returns a promise with the response.
-* `delete ( id [, headers ] )`: Delete a member in a collection. Returns a promise with the response.
+* `delete ( id [, headers, data ] )`: Delete a member in a collection. Returns a promise with the response.
```
Comment (jeromemacias) [inline README.md]: You should update the documentation according to your last commit:
`delete ( id [, data, headers ] )`

File: README.md [hunk 13]
```diff
@@ -222,7 +222,7 @@ authorsCollection.get(1).then(function(authorEntity) { /*  */ });
 
 * `get ( [ params [, headers ]] )`: Get a member. Returns a promise with an entity.
 * `put ( data [, headers ] )`: Update a member. Returns a promise with the response.
-* `delete ( [ headers ] )`: Delete a member. Returns a promise with the response.
+* `delete ( [ headers, data ] )`: Delete a member. Returns a promise with the response.
```
Comment (jeromemacias) [inline README.md]: Same here.

**Comment by jeromemacias** [comment]
:+1:

**Comment by bkotyik** [comment]
Oh, I forgot it sorry. No its fixed!

**Comment by fzaninotto** [comment]
Thanks!

**Pull request merged**